src/pushsource/_impl/model/vms.py

```diff
@@ -57,6 +57,11 @@ class VMIPushItem(PushItem):
     description = attr.ib(type=str, default=None, validator=instance_of_str)
     """A brief human-readable description of the image."""
 
+    uefi_support = attr.ib(
+        type=bool, default=None, validator=optional(instance_of(bool))
+    )
+    """``True`` if the image supports UEFI boot."""
+
```

Comment:
It can be boolean if there is another boolean as well to cover whether legacy BIOS is supported.  @amcmahon-rh had mentioned that on the other PR.  If your final comment there was meant to explain why that won't work, that wasn't clear (to me at least).

There's related work currently in draft at https://url.corp.redhat.com/89a87c2 which is aiming for two booleans:

- `bios_support`: image can be booted with (legacy) BIOS
- `uefi_support`: image can be booted with UEFI

...and it would then be up to the user of the push item to decide how to express this info to AWS or Azure APIs or any other layer which needs to know.

Is there any reason that wouldn't work, or reason that a single string attribute should be preferred?

Reply:
emulating enum with many boolean does not work for me;
it can result into invalid states e.g.
```
bios_support = False
uefi_support = False
```
or other combinations might be not very obvious but valid
```
bios_support = True
uefi_support = True
``` 
One might interpret that as "uefi-preferred". But people might be confused while reading the code.

The issue is you get 9 combinations with possible values (Cartesian product)
`{True, False, None #for bios_support} X {True, False, None #for uefi_support}`

One would not be able to achieve invalid states with enum(string)
and they would be more self descriptive.

@JAVGan looks like we cannot cannot change it for backward compatibility reason.
Please create the new field  uefi_support_enum
```
    uefi_support_enum = attr.ib(
        type=bool, default=None, validator=optional(instance_of(str))
    )
    """
        hybrid, uefi, legacy or  missing value which means AWS default
        more details https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ami-boot-mode.html
    """
```
or we might implement validator for subset of allowed values
*https://www.attrs.org/en/stable/api.html#attrs.validators.in_
```
attrs.validators.optional(attrs.validators.in_(["hybrid", "uefi", "legacy"]))
```

Reply:
I'm going to share the table I made for the PR that Rohan linked above to help clarify the point of using booleans.

In the future we may have to add similar bootmode enablement for other cloud platforms, which could use other string values. Keeping the values abstract until the point where the AWS / other cloud provider API is used saves us from locking ourselves in with specific string values. 


![Screenshot from 2023-08-30 10-18-40](https://github.com/release-engineering/pushsource/assets/100223330/48de6f3d-b320-4b52-8426-88a0cd6eb8ba)

Reply:
@amcmahon-rh your table is not sufficient. The file is optional and thus allow `None`.
which makes thigs more complicated
`bios_support = False && uefi_support = False` is **invalid** combination and it should raise an exception
but `bios_support = None && uefi_support = None` is **perfectly valid**.
Which means you do not want to set `BootMode` at all and you want to rely on AWS defaults (which was state few weeks ago)
And might be perfectly valid for some cases.

> In the future we may have to add similar bootmode enablement for other cloud platforms,

Bootmode with enum-like strings(or even enums directly) is more extensible than combination of  2 booleans or 3,4,5 booleans in the future.

Reply:
> @JAVGan looks like we cannot cannot change it for backward compatibility reason.
> Please create the new field uefi_support_enum

Even though I understand the `Enum` point it will create a redundant attribute for `VMIs` on 'AmiPushItem` and one can opt to use the boolean in favor of our `Enum` which isn't a good solution as well....

Ideally we should replace this or go with the 2nd boolean if there's no way on removing that in favor of `Enum`.

Adding a 2nd attribute which will be redundant is asking for implementation mess up IMHO

Reply:
We need to introduce another attribute anyway to support all required cases.
* Introducing enum/str would allow to deprecate current `uefi_support`
* Introducing new boolean will just make code more complicated. And will not be much extensible for other VM-types*(marketplaces) in the future.

Reply:
@lslebodn I've adjusted it, PTAL again

Reply:
tyvm
 +1

Reply:
I'm not convinced about some of these enum arguments.

About an enum being more extensible.  I agree with that when accepting a value *as input*, but here we are producing it *as output* and I think it might be the other way around.

Let's say `boot_mode` is an enum.  If a new value is ever added to the enum, 100% of existing users of this library won't know how to deal with it, right?

If the new value was supposed to express (say) UEFI support + "something else", then potentially it'd make sense for current users to treat the image as just UEFI until they're updated to care about the "something else".  But they'd have no way to do that when given a string they've never seen before.  Whereas, if they were coded against a `uefi_support` boolean which was still set to True, that code would still correctly understand that the image supports UEFI and could continue to work reasonably until updated to care about whatever separate field expresses the "something else" (if that even turns out to be necessary).

About booleans allowing an invalid combination of values, I'm not sure how we can say that those are certainly invalid?  In the case of `bios_support = False && uefi_support = False` , perhaps today you don't expect to feed any images into this where that'll be the case, but nothing seems intrinsically wrong about it.  VMIPushItem does not even say that it represents a "bootable" image and nothing seems to preclude using it for non-bootable images (e.g. an image to be attached as a secondary disk).  And if there's ever a successor to UEFI and we produce images which only support that, it would indeed be correct to set `bios_support = False && uefi_support = False` on those.

Ultimately I'm willing to merge either solution and I lean on the side of "those who do the work make the choices", so I'm not asking this to be changed.  I just wanted to provide some counterpoints because the arguments above in the thread seem very confidently in favor of enum, while in my opinion it's not so clear-cut and neither solution is obviously better than the other.

Reply:
>  If a new value is ever added to the enum, 100% of existing users of this library won't know how to deal with it, right?

Sure but that is expected with enum. You handle supported cases in your library
and remaining are either logged as a warning and use fallback
or better with defensive approach you fail(raise exception) with unsupported input value.

Current implementation is defensive due to validators
 
But log message/exception for enum is much simpler to understand than error
```
raise Exception(f"Unsupported combination of bool1:{bool1} bool2:{bool2} bool3:{bool3} bool4:{bool4}")
```
and than checking support table somewhere in documentation

>  In the case of bios_support = False && uefi_support = False
It needn't mean just non-bootable image. There are few interpretations:
* non-bootable images (e.g. an image to be attached as a secondary disk).
* bug in the code
* 3rd new option for booting image and creative developer

> Ultimately I'm willing to merge either solution and I lean on the side of "those who do the work make the choices"
@amcmahon-rh Does current state of MR works for you?

Reply:
I'll accept the Enum based approach here. I think it may require a little more care and maintenance for our tooling, and has some risks as Rohan pointed out, but the solution works.

Reply:
Would it be possible to use Enums specifically for the boot_mode type, rather than strings? I feel like Enums are a more rigorous approach than strings.